### DISCUSSION TITLE
Nixie x MEGA-BAS RS-485 Communication (Intellichlor40)

### DISCUSSION THREAD

**Discussion by craywm** [tagyoureit/nodejs-poolController#844] [RS485]
*Continuation of a previous #841 discussion. Please reference back to that discussion for additional back ground information.

1. I re-imaged the RPi with fresh image utilizing [this](https://github.com/rstrouse/nodejs-poolController-dashPanel/wiki/nodejs-poolController-dashPanel-setup-and-configuration-on-a-Raspberry-Pi-for-the-novice) guide. I only utilized for RPi setup (basic)

2. I then enabled VNC and headless resolution (remote function): 
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/27ca1d5c-ceb1-4b88-8643-b89785261a68)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/38084c3c-fe09-4e8b-b32f-aa6af5fd0225)

3. I decided to utilize [this](https://github.com/tagyoureit/nodejs-poolController/wiki/Quick-Start-Guide) guide to install the pre-requisites, REM, Poolcontroller and Dashboard. All the guides are pretty much the same, and I'm referencing here to simply point back to what was entered into the terminal for installation. I did also install pm2 and lograte and setup to start on boot. 

4. Setup Nixie with [this](https://github.com/tagyoureit/nodejs-poolController/wiki/DIY-Standalone-Nixie-Pool-Controller#configuring-the-software)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/7b32f98d-70e7-4b53-a0f1-f3897a2d4885)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/5057f099-e0be-4563-bc31-456fe973a39a)

5. Setup MEGA-BAS in REM using [this](https://github.com/rstrouse/relayEquipmentManager/wiki/Sequent-Microsystems-Boards#sequent-building-automation-hat-mega-bas) guide.
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/70b4b9dd-e0d1-4b1c-b94b-e864f609d707)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/babc5aed-c856-4ee1-95c4-db6bed460e7a)

just a note here, my 8-RELAYS HAT was not automatically identified:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/4c7b65f7-71d1-4c8f-a4dd-46df873976c6)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/34e16d76-8068-4571-8fb4-a2ad74d3036a)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/d0f33ad9-ddb0-48c1-849c-317544ba6a4a)

6. Input Long/Lat into Nixie, setup body / filter data, setup pump #1 (Just because).

7. Setup RS-485 Comms (more detailed):
Here is what my RPi config looks like:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/7e3826ea-c685-4087-9515-904897ae8ce6).
Turned on Serial Port:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/8c1de267-8993-441f-9665-9fca3c29d5b1)

At this point, I'm fae enough along in setup to utilize the RS-485 comms and confirm communication. Navigating to Comms tab showed by default (I did not change anything):
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/2edf2301-1adc-4be6-b2ee-38c3e2f09279)

I have not installed a USB device yet. Selected ttyS0 device:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/281d2939-f739-4d20-b45f-c6e5e391515b)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/425dd4e6-0fdc-4d9a-960f-174d87f8e5cf)

Navigated to Chemistry and setup a chlorinator:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/fe225881-ad88-4be4-9a5c-f1cf1e9c4fda)

Plugged everything in outside:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/3a184889-0142-4d9d-9a9f-0a6e43b8e492)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/8e045b9c-51d7-4ed4-befe-356b3e30887a)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/6469c4bc-cb92-4d5d-9e34-ccdc58d4582a)

I'm getting the same condition. I have not yet tried to plug in a USB to see if that will allow data to flow. It should be noted that I sent some data from the Mega-bas RS-485 port to my computer to confirm that communication was OK (and it was). I initially left the jumpers on the 8-RELAYS, but removed. I also tried to swap the wires with no success. The inbound bytes are being tossed due to some unrecoverable collision. It should also be noted that I am using no device overlays, and have not modified any config files. 

I can remove the 8-relays HAT and configyre. Also, I can set up with USB to see if packet by packet is the same, or if collisions are occuring on the Mega-bas side and not the USB side. does anyone have suggestions to move forward trouble shooting, or identified errors in my above setup?

**Comment by rstrouse** [comment]
The collision detection works like this.  The software is constantly monitoring the serial port for inbound data.  If nothing is communicating then no bytes will be read from it.  It should be silent.  Then when bytes come in the software looks for preamble bytes indicating an incoming message frame.  In your case this should be [16, 2].  If it does not find it in the max packet length then these bytes are determined to be noise and the buffered byte stream tosses those bytes waiting for a preamble.

Expand the pm2 monitor console and at the bottom in the metadata section you should see Node.js version.  Which version are you seeing there?

**Reply by craywm** [reply]
node.js version:  18.17.1. I think previous build used version 20. 

![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/8681a614-6cbe-4a14-b77d-57901c0603bd)

**Reply by rstrouse** [reply]
Stop njsPC on PM2 by running `pm2 stop njsPC`  then try running it using the following commands and test.

`cd ~/nodejs-poolController`
`npm start`

Leave this console window open and test the port.

**Reply by craywm** [reply]
Is this different from pm2 monit? I’ll try that when I get home. 

Is it possible that no ground reference is causing the IC40 rs-485 signal to be misinterpreted by the mega-bas?
I’ve thought about running a cable from ground on the surge protected to the ground on the mega-bas, but have read it’s unnecessary.

**Reply by craywm** [reply]
The communication in console is same as previously:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/0fd7f265-fac3-4c97-ab1c-b8604b2b9562)
 
Port is also same condition:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/47279062-346b-412e-9e77-b0005206c676)

These are the variations of the collision messages:
[8/29/2023, 6:55:27 PM] verbose: Tossed Inbound Bytes 25,223,249,252,183,95,221,255,25,223,249,248,183,95,221,255,25,223,249,252 due to an unrecoverable collision.
[8/29/2023, 6:55:33 PM] verbose: Tossed Inbound Bytes 59,223,249,252,183,95,255,255,59,223,249,252,183,95,221,255,25,223,249,252 due to an unrecoverable collision.
[8/29/2023, 6:55:30 PM] verbose: Tossed Inbound Bytes 183,95,255,255,59,223,249,248,183,95,255,255,59,223,249,252,183,95,255,255 due to an unrecoverable collision.
[8/29/2023, 6:55:17 PM] verbose: Tossed Inbound Bytes 183,95,221,255,25,223,249,252,183,95,221,255,25,223,249,252,183,95,255,255 due to an unrecoverable collision.
[8/29/2023, 6:55:05 PM] verbose: Tossed Inbound Bytes 25,223,249,248,183,95,221,255,25,223,249,252,183,95,255,255,59,223,249,248 due to an unrecoverable collision.

**Comment by CodeSlinger69** [comment]
Have you tried testing with a different RS485 device other than the chlorinator? Do you have a pump you can connect to the bus? Just trying to eliminate possibilities.

**Reply by craywm** [reply]
I tested the RS-485 communication to both the 8-Relays and Mega-bas to using the USB adapter plugged into my computer running Realterm. I was able to receive and send data through both boards, and simulate sending the status command to njsPC which did push status correctly to dashpanel. All is good on that side. Plugged it back up to surge protector at pad, and same story. 

I’m thinking of plugging the SWC surge protected A/B wires to my usb adapter on my laptop and see what realterm comms look like. Only issue here is that the connection to the USB adapter has been successful previously.

To answer your question directly, I don’t have any other RS-485 devices, but feel that my confirmation above should suffice. 

Is it possible to push the surge protector board into some other “mode”? It’s odd that all the bits are ignored. Could the baud rate on the SPB have been changed somehow?

**Reply by craywm** [reply]
As it turns out, my plans changed for this week and I now have access to another pool with a pentair inteliflo vs pump and an IC40. 

I plan to try the IC40 first then the pentair pump with RS-485 comms. 

At the original pool, I continued to receive tossed packets through the mega-bas. I plugged in the USB adapter and wired to the bus, and without setting up any device or chlorinator, data began to flow through port0 on device ttyS0 same as before. Communication was perfect. When I removed the USB, it immediately stopped. 

Hopefully my testing here will be more successful and aid to narrow issue. Is there some driver for the Mega-bad that should be running? It’s odd that when I plug in the usb, all other devices seem to start working even with no data flowing through the adapter.

**Comment by craywm** [comment]
I had some time last night to attach reconfigure and hook up RS-485 from a different IC-40 to my pool controller:
The Power center is a newer style board than the previous one I tested with
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/571ff335-edc0-4724-94c0-94f2f1c44b17)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/2363a77a-d681-49fb-a3f6-7999439fb665)

Fired everything up hoping for the best:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/8a51df83-e34e-4a5d-a856-6ba3a8798541)

Same issue as before. I'm going to try the communication on a Pentair Intelliflo VS pump today, but I expect the same result. 
I guess its a good thing that its not my equipment, but this is becoming frustrating. Is it possible Sequent changed something on the Mega-bas board causing this issue?

**Reply by rstrouse** [reply]
Did you try the port on the 8-relay card with the port disabled on the MEGA-BAS.  Perhaps look carefully at all the solder joints on the MEGA-BAS.  Does this I2c pin look like it jumped out of the way of the solder iron.

![image](https://github.com/tagyoureit/nodejs-poolController/assets/47839015/2782f2dd-e369-4f9f-9f71-880ced2f5fd8)

**Reply by craywm** [reply]
hmm.. thats a good observation rstrouse. I'll check, but I do believe my I2C connection is good. I can hooked up a thermistor and did not see any irregularities. I will however double check! 
I decided to verify that my GPIO were being pulled up/down and:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/fb5886b4-00bd-4223-a161-53e60e783e32)
shouldn't GPIO 14 be pulled high/low?

Also checked dmesg for reference to serial and tty:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/74ab36aa-5c18-43ce-9346-6828bf8e63c3)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/6a1b1f22-abfc-4105-ad88-c7456da2278a)

**Reply by craywm** [reply]
The I2C pins are OK (there is some solder on the underside of the board, could be better. 
But your comment about the pins made me think to check the GPIO pins as well:
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/906b9896-cc26-4da0-a586-b4806222715a)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/4f97197f-b4c0-4211-a0cd-815f94e7e0c3)
![image](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/89171bd8-3720-45b0-947d-2f7efb519362)
I'm wondering if it is shorting to ground right across (pin 9). This would make sense as to why the I require two input paths to the serial port (Transmitting/receiving on different devices). 

I would expect the 8-Relays card to have the same issue since it is stacked on top of the Mega-bas, but the 8-relays card actually only has a jumper to connect the tx side of the board to the GPIO pins, the RX side is always connected (might have reversed that). Thus when I connected both board to the bus, communication was successful. 

I'm going to run a GPIO diagnostic to make sure everything is OK on the pi side, then do some continuity tests on that pin. I'm not near my soldering equipment now, so I cant fix it today (if that is the issue), But I should be able to stack just the 8-relays card ontop of the pi to confirm that is the issue (again assuming the pi GPIO is OK).

**Reply by craywm** [reply]
GPIO pins checked out OK. Checked continuity on mega-bas board between pins 10 (GPIO 15) and 9 (GND) with no continuity. I’m going to check with the mega-bas card stacked as well (no power) and with power. 

I went ahead and stacked the 8-Relays on the rpi, installed jumpers, and fired everything up. 
![IMG_0808](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/d429c950-9678-4c99-9a39-2bd92a9ed1f5)
 And the verdict….
![IMG_0809](https://github.com/tagyoureit/nodejs-poolController/assets/18619654/44afb278-c092-4335-924b-e8c9f7bc1150)

SUCCESS!! As I suspected, the Pi side was ruled out. I’m still thinking there’s a short on the Mega-bad board causing the 8-relays to also short when all stacked. To test I’ll follow this sequence:
1. Restock HAT pi->BAS->8RLY 
2. No jumpers on BAS
3. Jumpers installed on 8-Relay
4. Test pin continuity to GND without power
5. If short exists, should identify at GPIO pins to 8relay level. 
6. power on and check continuity to GND
7. If no short, then removed jumpers on BAS have isolated the anomaly
8. Test RS-485 pass through on 8-relays. 
9. Reach out to Sequent to request support/replacement. 

Has there been any history of defective Mega-BAS cards to your knowledge?

**Reply by rstrouse** [reply]
I have stacked up several BAS cards and never had a problem.  The symptoms are so very odd.

**Reply by CodeSlinger69** [reply]
At this point I would just get a replacement from whatever vendor you bought this from. You've done far more work than I would have done in trying to track down the problem, especially after noticing that wonky soldering on the board.

**Comment by jamestalmage** [comment]
I am experiencing the identical symptoms. I'm receiving packets, but 100% transmit errors.

@craywm Did you find a resolution?

**Reply by jamestalmage** [reply]
I notice @craywm is using a version 4.0 board, same as me.

How old are the 3.X boards? Has anyone had success with the new 4.X boards yet? Wondering if there's something to the guess that something is wrong/changed with the new boards.

Is there a command that would let me just dump the received bits in a way that would make sense to post? At least verify the expected preamble is being received?

I tried the `screen` command... Just black characters... Guessing they don't translate to ASCI

**Reply by tagyoureit** [reply]
A good first step is to try and [update the firmware](https://github.com/SequentMicrosystems/megabas-rpi/blob/master/update/README.md).  When you do this, make sure that none of the applications accessing the card (njsPC, REM, etc) are running).

Have you tried swapping the A/B cables?  

You can dump the raw bits [with od -x](https://github.com/tagyoureit/nodejs-poolController/wiki/RS-485-Adapter-Details#maclinux), but [MessageManager](https://github.com/rstrouse/nodejs-poolController-dashPanel/wiki/Message-Manager) in dashPanel is much more effective.

**Reply by jamestalmage** [reply]
@tagyoureit I'm on firmware 2.03, and yes. Definitely swapped A/B. I'll head out and reconnect and send the MessageManager info in a bit

**Comment by jamestalmage** [comment]
@tagyoureit I tried everything you suggested:

The comms page shows bytes are being received:


<img width="645" alt="Screenshot 2023-09-14 at 10 05 56 PM" src="https://github.com/tagyoureit/nodejs-poolController/assets/4082216/ee1d6a93-14ad-4f19-9bf0-6a0af8cb760c">


But Message Manager only shows outgoing messages:

<img width="586" alt="Screenshot 2023-09-14 at 10 16 25 PM" src="https://github.com/tagyoureit/nodejs-poolController/assets/4082216/2c155e10-09c9-4d7d-94cc-436c4ac9e4d3">


```
ox -d /dev/ttyS0
0000000 0100 3f2d f1bd fdff 00a5 0100 3f2d f1bd
0000020 fdff 00a5 0100 3f2d f1bd fdff 00a5 0100
0000040 3f2d f7bd 01fd 35bf 0100 3f2d f7bd 01fd
0000060 35bf 0100 3f2d f3bd f7fd 9dfd 0000 2d01
0000100 bd3f fff1 a5fd 0000 2d01 bd3f fff1 a5fd
0000120 0000 2d01 bd3f fff1 a5fd 0000 2d01 bd3f
0000140 fdf7 bf01 0035 2d01 bd3f fdf7 bf01 0035
0000160 2d01 bd3f fdf3 fdf7 009d 0100 3f2d f3bd
0000200 f7fd 9dfd 0000 2d01 bd3f fff1 a5fd 0000
0000220 2d01 bd3f fff1 a5fd 0000 2d01 bd3f fff1
0000240 a5fd 0000 2d01 bd3f fdf7 bf01 0035 2d01
0000260 bd3f fdf7 bf01 0035 2d01 bd3f fdf3 fdf7
0000300 009d 0100 3f2d f3bd f7fd 9dfd 0000 2d01
```

The `ox -d` command shows nothing if the `nodejspoolcontroller is not running.

A USB adapter arrived today, going to use it to verify wiring tomorrow while the sun is shining

**Comment by craywm** [comment]
@jamestalmage i ended up going with the RS-485 on the 8-Relays. You should not have an issue with the USB adapter if it is the same issue as mine. I do think there’s some interference on the board causing the issue. I was able to send to my PC through the megabas, but not receive, data would come, but it was corrupted. 

Change your log settings to debug and check PC2 monit for njsPC tossing data packets incoming. This should indicate that data is reaching the pi from through the megabas RS-485 transceiver, however it is corrupted. 

Just FYI, if you go with the 8relays transceiver (assuming you have one), remove jumpers from the Mega-BAs RS-485 Tx/Rx pins (as mentioned above) to isolate the Megabas from the GPIO pins. This is where I spun my wheels for a bit, removed jumpers and smooth sailing with 8-Relays.

**Comment by jamestalmage** [comment]
@rstrouse @craywm Turns out my issue was miswiring on the pump end. Now to get the AquaCal heater working.

**Comment by mhmatttu** [comment]
I have experienced a similar issue with my mega-bas v4.  I have an intelliflo vsf pump and ic40.  when either are connected to the mega-bas rs485, they work fine with zero failure rate.  However, when i connect both of them at the same time, i get 100% failure send rate. I swapped wires, ensured good connections, added enable_uart=1 to boot config, rs485 pass through enabled, verified correct rpi interface settings.  I currently have the pump connected to the mega-bas and the ic40 connected to a USB adapter and all is working well.  Any thoughts on why both will not work together on the mega-bas?

![IMG_0360](https://github.com/tagyoureit/nodejs-poolController/assets/47801991/565c37c1-2d4c-4a32-8b90-1857cdae3b7c)

**Reply by craywm** [reply]
Sounds to me like the address for the IC40 and Intelliflo are overlapping. Sometimes you have to manually set the address (need to do some ping to boards to confirm the address, or usually the default in the manuals is correct unless Sequent changed them again on newer boards).  
Can you take a screen shot of each REM setting when you setup as Mega-bas passthrough?
Just a note, this issue is different than mine, as my Mega-BAS board was shorting across the receiving RS485 (physical board issue). I did reach out to Sequent for a replacement but never received a response. I went for the 8-relays hat pass through instead and have never looked back. 
Also one more note, its been a while since I've changed anything, so I'm a bit rusty on troubleshooting and current developer issues. So please someone chime in if I'm way off with my response.

**Reply by mhmatttu** [reply]
Thanks.  Yeah, my mega-bas rs485 clearly seems to be working but for some reason it is not allowing the bus topology (i think that is what it’s called) where multiple devices can be connected.  Attached are some screenshots.

![IMG_0365](https://github.com/tagyoureit/nodejs-poolController/assets/47801991/e0a52f1a-f02c-4e16-9ae5-2ee530494201)
![IMG_0364](https://github.com/tagyoureit/nodejs-poolController/assets/47801991/b7f1a666-03dd-4a15-a4f2-e7d2607d5e1b)

**Reply by rstrouse** [reply]
Actually the one difference with the port on the BAS is that it is wired for modbus support.   Try disabling it with the cli.  While I love the Sequent cards it is utterly frustrating that they don’t document register changes on the I2c and they make indiscriminate changes as if nobody uses the hardware for anything other than playing with it.  Heck they don’t document the registers at all.  You have to manually decrypt c enumerated to come up with the values.

**Reply by mhmatttu** [reply]
Thanks.  Apologies for my ignorance but how do I disable it?  Below are the settings I get with megabas 0 rs485rd

![IMG_0366](https://github.com/tagyoureit/nodejs-poolController/assets/47801991/05da3b5f-0e54-479c-a09a-9063ee3489e2)

**Reply by craywm** [reply]
> Actually the one difference with the port on the BAS is that it is wired for modbus support. Try disabling it with the cli. While I love the Sequent cards it is utterly frustrating that they don’t document register changes on the I2c and they make indiscriminate changes as if nobody uses the hardware for anything other than playing with it. Heck they don’t document the registers at all. You have to manually decrypt c enumerated to come up with the values.

Thanks @rstrouse. This is the condition I attempted to describe above. Hope this fixes your problem @mhmatttu!

**Reply by mhmatttu** [reply]
Based on the below statement from the sequent website, should I change the rs485 settings to
 `megabas [0]  rs485wr 0 0 0 0 0`?

![IMG_0367](https://github.com/tagyoureit/nodejs-poolController/assets/47801991/c61b03c1-e0e3-4bd6-83f0-cdf0067971c0)

**Reply by mhmatttu** [reply]
By way of follow-up, the above command did resolve my issue.  I now have the intelliflo and IC40 connected to the Mega-bas RS485 port and working.
